core/src/main/java/io/grpc/util/GracefulSwitchLoadBalancer.java

```diff
@@ -0,0 +1,124 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.util;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.ConnectivityState.READY;
+
+import io.grpc.ConnectivityState;
+import io.grpc.Internal;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancerProvider;
+import io.grpc.Status;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A load balancer that gracefully swaps to a new lb policy. If the channel is currently in a state
+ * other than READY, the new policy will be swapped into place immediately.  Otherwise, the channel
+ * will keep using the old policy until the new policy reports READY or the old policy exits READY.
+ */
+@Internal
+@NotThreadSafe // Must be accessed in SynchronizationContext
+public final class GracefulSwitchLoadBalancer extends ForwardingLoadBalancer {
+  private static final LoadBalancer NOOP_BALANCER = new LoadBalancer() {
+    @Override
+    public void handleNameResolutionError(Status error) {}
+
+    @Override
+    public void shutdown() {}
+  };
+
+  private final Helper helper;
+  private LoadBalancer currentLb = NOOP_BALANCER;
+  private LoadBalancer pendingLb = NOOP_BALANCER;
+  @Nullable
+  private String currentPolicyName;
+  @Nullable
+  private String pendingPolicyName;
+  private boolean isReady;
+
+  public GracefulSwitchLoadBalancer(Helper helper) {
+    this.helper = checkNotNull(helper, "helper");
+  }
+
+  /** Gracefully switch to a new load balancing policy. */
+  public void switchTo(LoadBalancerProvider newLbProvider) {
+    checkNotNull(newLbProvider, "newLbProvider");
+
+    String newPolicyName = newLbProvider.getPolicyName();
+    if (newPolicyName.equals(pendingPolicyName)) {
+      return;
+    }
+    pendingLb.shutdown();
+    pendingLb = NOOP_BALANCER;
+    pendingPolicyName = null;
+    if (newPolicyName.equals(currentPolicyName)) {
+      return;
+    }
+
+    class PendingHelper extends ForwardingLoadBalancerHelper {
+      LoadBalancer lb;
+
+      @Override
+      protected Helper delegate() {
+        return helper;
+      }
+
+      @Override
+      public void updateBalancingState(ConnectivityState newState, SubchannelPicker newPicker) {
+        if (pendingLb == lb && newState == READY) {
+          swap();
+        }
+        if (currentLb == lb) {
+          helper.updateBalancingState(newState, newPicker);
+          isReady = newState == READY;
+          if (!isReady && pendingLb != NOOP_BALANCER) { // current policy exits READY, swap
```

Comment:
When the old policy is swapped out due to non-readiness, the last reported state from the new policy should take over. This will make a difference on fail-fast RPCs if old policy is TRANSIENT_FAILURE while the new one is CONNECTING. How about:
```java
if (pendingLb == lb) {
  pendingState = newState;
  pendingPicker = newPicker;
  if (newState == READY) {
    swap();
  }
}
if (currentLb == lb) {
  isReady = newState == READY;
  if (!isReady && pendingLb != NOOP_BALANCER) {
    swap();
    helper.updateBalancingState(pendingState, pendingPicker);
  } else {
    helper.updateBalancingState(newState, newPicker);
  }
}
```

Reply:
What is the initial pendingPicker?

A. null, on swap helper will not update picker for new policy.
B. BufferPicker, on swap helper will update BufferPicker for new policy.

Reply:
The initial pendingPicker is relevant only when `currentLb` is swapped out before `pendingLb` emits an update. In that case initial state should be IDLE and the picker should be BufferPicker, which is consistent with the initial states of ManagedChannel before LoadBalancer emits the first update.

Reply:
Done, with a little difference.